db/active_replicator_common.go

```diff
@@ -24,15 +25,154 @@ type activeReplicatorCommon struct {
 	replicationStats      *BlipSyncStats
 	onReplicatorComplete  ReplicatorCompleteFunc
 	lock                  sync.RWMutex
+	doReconnect           chan struct{}
+	ctx                   context.Context
+	ctxCancel             context.CancelFunc
+}
+
+func newActiveReplicatorCommon(config *ActiveReplicatorConfig, direction ActiveReplicatorDirection) *activeReplicatorCommon {
+
+	var replicationStats *BlipSyncStats
+	if direction == ActiveReplicatorTypePush {
+		replicationStats = BlipSyncStatsForSGRPush(config.ReplicationStatsMap)
+	} else {
+		replicationStats = BlipSyncStatsForSGRPull(config.ReplicationStatsMap)
+	}
+
+	logCtx := context.WithValue(context.Background(), base.LogContextKey{}, base.LogContext{CorrelationID: config.ID + "-" + string(direction)})
+	ctx, ctxCancel := context.WithCancel(logCtx)
+
+	return &activeReplicatorCommon{
+		config:           config,
+		state:            ReplicationStateStopped,
+		doReconnect:      make(chan struct{}),
```

Comment:
Review potential deadlocks associated with multiple sends to this channel while reconnect retryloop is running.